sbncode/Calibration/TrackCaloSkimmer_module.cc

```diff
@@ -0,0 +1,460 @@
+  fTrack->ndaughters = pfp.Daughters().size();
+
+  fTrack->length = track.Length();
+  fTrack->start_x = track.Start().X();
+  fTrack->start_y = track.Start().Y();
+  fTrack->start_z = track.Start().Z();
+  fTrack->end_x = track.End().X();
+  fTrack->end_y = track.End().Y();
+  fTrack->end_z = track.End().Z();
+  fTrack->dir_x = track.StartDirection().X();
+  fTrack->dir_y = track.StartDirection().Y();
+  fTrack->dir_z = track.StartDirection().Z();
+
+  // Fill each hit
+  for (unsigned i_hit = 0; i_hit < hits.size(); i_hit++) {
+    sbn::HitInfo hinfo = MakeHit(*hits[i_hit], hits[i_hit].key(), *thms[i_hit], track, calo, rawdigits);
+    if (hinfo.plane == 0) {
+      fTrack->hits0.push_back(hinfo);
+    }
+    else if (hinfo.plane == 1) {
+      fTrack->hits1.push_back(hinfo);
+    } 
+    else if (hinfo.plane == 2) {
+      fTrack->hits2.push_back(hinfo);
+    }
+  }
+
+  // Hit summary info
+  fTrack->hit_min_time_p0_tpcE = HitMinTime(fTrack->hits0, true);
+  fTrack->hit_max_time_p0_tpcE = HitMaxTime(fTrack->hits0, true);
+  fTrack->hit_min_time_p0_tpcW = HitMinTime(fTrack->hits0, false);
+  fTrack->hit_max_time_p0_tpcW = HitMaxTime(fTrack->hits0, false);
+  fTrack->hit_min_time_p1_tpcE = HitMinTime(fTrack->hits1, true);
+  fTrack->hit_max_time_p1_tpcE = HitMaxTime(fTrack->hits1, true);
+  fTrack->hit_min_time_p1_tpcW = HitMinTime(fTrack->hits1, false);
+  fTrack->hit_max_time_p1_tpcW = HitMaxTime(fTrack->hits1, false);
+  fTrack->hit_min_time_p2_tpcE = HitMinTime(fTrack->hits2, true);
+  fTrack->hit_max_time_p2_tpcE = HitMaxTime(fTrack->hits2, true);
+  fTrack->hit_min_time_p2_tpcW = HitMinTime(fTrack->hits2, false);
+  fTrack->hit_max_time_p2_tpcW = HitMaxTime(fTrack->hits2, false);
+
+  // Save information on a fit to the end of the track
+  if (fDoTailFit) DoTailFit();
+
+  // get information on nearby tracks
+  for (const GlobalTrackInfo &othr: tracks) {
+    if ((track.End() - othr.start).r() < 50. || (track.End() - othr.end).r() < 50.) {
+      fTrack->tracks_near_end_dist.push_back(std::min((track.End() - othr.start).r(), (track.End() - othr.end).r())); 
+      fTrack->tracks_near_end_costh.push_back(
+        (track.End() - othr.start).r() < (track.End() - othr.end).r() ? 
+          track.EndDirection().Dot(othr.dir) : track.EndDirection().Dot(othr.enddir)); 
+    }
+  }
+
+  // Save?
+  bool select = false;
+  if (!fSelectionTools.size()) select = true;
+
+  // Take the OR of each selection tool
+  int i_select = 0;
+  for (const std::unique_ptr<sbn::ITCSSelectionTool> &t: fSelectionTools) {
+    if (t->Select(*fTrack)) {
+      select = true;
+      fTrack->selected = i_select;
+      break;
+    }
+    i_select ++;
+  }
+  
+  // Save!
+  if (select) {
+    std::cout << "Track Selected!\n";
+    fTree->Fill();
+  }
+}
+
+void sbn::TrackCaloSkimmer::DoTailFit() {
+  // Try fitting the constant and exponentials to the tail of dQ/dx v. RR on the collection plane
+  std::vector<double> fit_rr;
```

Comment:
not sure how big these are, but it may be better to reserve for the total number of hits, so that the push_backs are a bit faster.

Reply:
It shouldn't be more than 10-15 points.

Reply:
OK. Given it may be doing this frequently, I'd suggest that maybe this be part of the class and just cleared after use ... to avoid allocations in loops and the like. But, if performance is good and it's not worth it, it's fine.

Reply:
I think the performance overhead of this step is minimal.